File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/TeleportStart.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;

/**
 * TeleportStart sim->viewer
 * announce a successful teleport request to the viewer.
 *
 * <p>Template: {@code TeleportStart Low 73 Trusted Unencoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code process_teleport_start()} in indra/newview/llviewermessage.cpp
 * (secondlife/viewer @ c179f76c01).
 */
public class TeleportStart extends SLMessage {
    public Info Info_Field;

    /** Block Info, Single. */
    public static class Info {
        public int TeleportFlags; // U32
    }

    public TeleportStart() {
        this.zeroCoded = false;
        this.Info_Field = new Info();
    }

    @Override
    public int CalcPayloadSize() {
        return 8;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleTeleportStart(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 73 (TeleportStart).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x49);
        packInt(byteBuffer, this.Info_Field.TeleportFlags);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.Info_Field.TeleportFlags = unpackInt(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/TeleportStart.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer

/**
 * TeleportStart sim->viewer
 * announce a successful teleport request to the viewer.
 *
 * <p>Template: {@code TeleportStart Low 73 Trusted Unencoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code process_teleport_start()} in indra/newview/llviewermessage.cpp
 * (secondlife/viewer @ c179f76c01).
 */
open class TeleportStart : SLMessage() {
    @JvmField var Info_Field: Info = Info()

    /** Block Info, Single. */
    open class Info {
        @JvmField var TeleportFlags: Int = 0
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return 8
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleTeleportStart(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 73 (TeleportStart).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x49).toByte())
        packInt(byteBuffer, Info_Field.TeleportFlags)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        Info_Field.TeleportFlags = unpackInt(byteBuffer)
    }
}
